### PR TITLE
Add option to use elogind

### LINKED ISSUE #226: meson `-Dwith-systemd=false` is ignored
**Describe the bug**
Regardless of whether `--Dwith-systemd` is set false or true, it detects systemd:
```
# meson setup --buildtype plain --prefix /usr/local --wrap-mode nofallback -Dwith-systemd=false build
[...]
Found pkg-config: /usr/bin/pkg-config (0.29.2)
Run-time dependency libsystemd found: YES 244
[...]
```

On a non-systemd system, this leads to the following error because systemd support seems to be forced:
```
 * abi_x86_32.x86: running multilib-minimal_abi_src_configure
meson setup --buildtype plain --libdir lib --localstatedir /var/lib --prefix /usr --sysconfdir /etc --wrap-mode nodownload --cross-file /tmp/portage/games-util/gamemode-1.5.1/temp/meson.i686-pc-linux-gnu.x86 -Dwith-systemd=false -Dwith-systemd-user-unit-dir=/usr/lib/systemd/user -Dwith-examples=false -Dwith-daemon=false /tmp/portage/games-util/gamemode-1.5.1/work/gamemode-1.5.1 /tmp/portage/games-util/gamemode-1.5.1/work/gamemode-1.5.1-abi_x86_32.x86
...
...
Run-time dependency libsystemd found: NO (tried pkgconfig and cmake)
meson.build:83:0: ERROR: Dependency "libsystemd" not found, tried pkgconfig and cmake
```

See-also: https://github.com/gentoo/gentoo/pull/13158#discussion_r411892796

### PR DESCRIPTION
**Pull request by stephanlachnit**
Closes #226.

Didn't test it yet.

### CODE REVIEW

**Comment by kakra** [comment]
Compiles fine for systemd with the patch applied, still waiting for elogind results...

**Comment by kakra** [comment]
Seems still to complain with the patch applied to 1.5.1:
```
>>> Configuring source in /tmp/portage/games-util/gamemode-1.5.1/work/gamemode-1.5.1 ...
 * abi_x86_32.x86: running multilib-minimal_abi_src_configure
meson setup --buildtype plain --libdir lib --localstatedir /var/lib --prefix /usr --sysconfdir /etc --wrap-mode nodownload --pkg-config-path=/usr/share/pkgconfig --cross-file /tmp/portage/games-util/gamemode-1.5.1/temp/meson.i686-pc-linux-gnu.x86 -Dsystemd=false -Delogind=true -Dwith-systemd-user-unit-dir=/usr/lib/systemd/user -Dwith-examples=false -Dwith-daemon=false --wrap-mode forcefallback /tmp/portage/games-util/gamemode-1.5.1/work/gamemode-1.5.1 /tmp/portage/games-util/gamemode-1.5.1/work/gamemode-1.5.1-abi_x86_32.x86
...
...
...
Run-time dependency libsystemd found: NO (tried pkgconfig and cmake)

meson.build:90:4: ERROR: Dependency "libsystemd" not found, tried pkgconfig and cmake
```

The meson cmdline has:

    -Dsystemd=false -Delogind=true -Dwith-systemd-user-unit-dir=/usr/lib/systemd/user

**Comment by stephanlachnit** [comment]
Typo:
`-Dwith-systemd=false -Dwith-elogind=true`

**Comment by kakra** [comment]
Oh, true... Gentoo doesn't add "with-" prefix to the feature name automatically...

**Comment by kakra** [comment]
LGTM, works in theory but there's a multilib problem somewhere else. But that should be unrelated to this PR. So from my point of view it's ready for merge.

**Comment by stephanlachnit** [comment]
> LGTM, works in theory but there's a multilib problem somewhere else. But that should be unrelated to this PR. So from my point of view it's ready for merge.

Same problem on Debian. It's not implemented yet properly, but we already split gamemode in 4 packages. This way the libraries are separated from the non-library files. If you want to have a look, [here](https://salsa.debian.org/stephanlachnit/gamemode_/-/tree/debian/master/debian) are Debian build files, the `xyz.install` files contain what is included in the package `xyz`.

**Comment by stephanlachnit** [comment]
Also I noticed that this isn't the best approach in Meson to solve this. Instead it should be a `type : 'combo'`, where one can choose between `['systemd', 'elogind', 'no-dameon']`, replacing the `with-daemon` option as well. If I have time I will change it.

**Comment by kakra** [comment]
Yes, a `no-daemon` option would be useful...

**Comment by stephanlachnit** [comment]
Pushed as combo.

**Comment by kakra** [comment]
Seems to close #222 for me: The abi32 compile pass no longer install the inih headers. Was this intentional?

**Comment by stephanlachnit** [comment]
It wasn't intentional, so no idea

**Comment by kakra** [comment]
That additional patch works for me. I'm good with it.

**Comment by stephanlachnit** [comment]
Can you send me the build log?

**Comment by kakra** [comment]
> Can you send me the build log?

Here's the complete log of the Gentoo ebuild, it builds for both abi64 and abi32, and it builds against systemd in my case:

[build.log](https://github.com/FeralInteractive/gamemode/files/4541805/build.log)

**Comment by kakra** [comment]
Looking at the build log, there should probably be an option to build without the libexec tools but that should be a matter of another issue report. The build installs the 64 bit versions over the 32 bit versions, so all is fine in the final package image. Just a few wasted CPU cycles. In Gentoo, packages are not split between ABIs, they build and install both in the same package if asked to do so. The build sources for this package are here if you're interested: https://github.com/gentoo/gentoo/pull/13158

**Comment by stephanlachnit** [comment]
> Here's the complete log of the Gentoo ebuild, it builds for both abi64 and abi32, and it builds against systemd in my case:

Something is wrong: `-Dwith-sd-bus-provider=systemd -Dwith-systemd-user-unit-dir=/usr/lib/systemd/user -Dwith-examples=false -Dwith-sd-bus-provider=no-daemon`

It's twice the same setting, and last one overrides, so the daemon isn't build at all, which means inih isn't builded at all.

**Comment by kakra** [comment]
> It's twice the same setting, and last one overrides, so the daemon isn't build at all, which means inih isn't builded at all.

Yes but that's expected behavior in this case: When building multilib, it adds this is the abi32 case. So it's good it overwrites the first setting.

From the ebuild:

```bash
multilib_src_configure() {
        local emesonargs=(
                -Dwith-sd-bus-provider=$(usex systemd "systemd" "elogind")
                -Dwith-systemd-user-unit-dir="$(systemd_get_userunitdir)"
        )
        if ! multilib_is_native_abi; then
                emesonargs+=(
                        -Dwith-examples=false
                        -Dwith-sd-bus-provider=no-daemon
                )
        fi

        # Upstream bug: https://github.com/FeralInteractive/gamemode/issues/222
        emesonargs+=(
                --wrap-mode forcefallback
        )

        meson_src_configure
}
```

So if building for abi32 only (and thus it would be the native abi), those extra args are not added.

**Comment by stephanlachnit** [comment]
Ah okay

**Comment by kakra** [comment]
> Ah okay

But neither abi pass install inih headers now. So that's probably a side-effect of using forcefallback. I didn't closely watch the behavior change between my ebuild adjustments.

**Comment by kakra** [comment]
There's a bug somewhere, from the build log:
```
 * abi_x86_64.amd64: running multilib-minimal_abi_src_configure
meson setup --buildtype plain --libdir lib64 --localstatedir /var/lib --prefix /usr --sysconfdir /etc --wrap-mode nodownload --pkg-config-path=/usr/lib64/pkgconfig -Dwith-sd-bus-provider=systemd -Dwith-systemd-user-unit-dir=/usr/lib/systemd/user --wrap-mode forcefallback /var/tmp/portage/games-util/gamemode-1.5.1/work/gamemode-1.5.1 /var/tmp/portage/games-util/gamemode-1.5.1/work/gamemode-1.5.1-abi_x86_64.amd64
```

In 64 bit build phase, it has `-Dwith-sd-bus-provider=systemd -Dwith-systemd-user-unit-dir=/usr/lib/systemd/user --wrap-mode forcefallback` and if I'm not blind, there's no other interfering option. But it doesn't build the daemon. It should, shouldn't it?

**Comment by kakra** [comment]
Tried again in my local dev build:
```bash
git clean -xdf && meson build && ninja -C build
```

There's no daemon in the build directory.

**Comment by stephanlachnit** [comment]
Maybe I messed something up, I'll look tmw.

File: meson.build [hunk 1]
```diff
@@ -168,7 +162,7 @@ endif
 
 # The daemon can be disabled if necessary, allowing multilib builds of the
 # main library
-if with_daemon == true
+if sd_bus_provider == 'no-dameon'
```
Comment (kakra) [inline meson.build:165]: This reverses the logic, doesn't it?

Comment (stephanlachnit) [reply meson.build:165]: Ah yes thanks.

**Comment by kakra** [comment]
Excerpt from the build log:
```
Run-time dependency dbus-1 found: YES 1.12.16
Run-time dependency threads found: YES
Library dl found: YES
Checking for function "pidfd_open" : NO
Configuring build-config.h using configuration
Looking for a fallback subproject for the dependency inih because:
Use of fallback dependencies is forced.

|Executing subproject inih method meson
|
|Project name: inih
|Project version: 48
|C compiler for the build machine: x86_64-pc-linux-gnu-gcc -m32 (gcc 9.3.0 "x86_64-pc-linux-gnu-gcc (Gentoo 9.3.0 p2) 9.3.0")
|C linker for the build machine: GNU ld.gold 2.33.1
|Appending CXXFLAGS from environment: '-O2 -march=native -pipe -fomit-frame-pointer -g -ftree-vectorize -ftree-slp-vectorize'
|Appending LDFLAGS from environment: ['-Wl,-O1', '-Wl,--as-needed']
|C++ compiler for the build machine: x86_64-pc-linux-gnu-g++ -m32 (gcc 9.3.0 "x86_64-pc-linux-gnu-g++ (Gentoo 9.3.0 p2) 9.3.0")
|C++ linker for the build machine: GNU ld.gold 2.33.1
|C compiler for the host machine: x86_64-pc-linux-gnu-gcc -m32 (gcc 9.3.0 "x86_64-pc-linux-gnu-gcc (Gentoo 9.3.0 p2) 9.3.0")
|C linker for the host machine: GNU ld.gold 2.33.1
|C++ compiler for the host machine: x86_64-pc-linux-gnu-g++ -m32 (gcc 9.3.0 "x86_64-pc-linux-gnu-g++ (Gentoo 9.3.0 p2) 9.3.0")
|C++ linker for the host machine: GNU ld.gold 2.33.1
|Build targets in project: 8
|Subproject inih finished.

Dependency inih from subproject subprojects/inih found: YES

data/meson.build:13:3: ERROR: Unknown variable "with_systemd".
```

**Comment by stephanlachnit** [comment]
Tbh I'm not that happy with combining everything into one option anymore, unless I would rewrite the entire build files so it makes a little bit more sense with this option.

**Comment by kakra** [comment]
Sorry:
```
data/meson.build:13:3: ERROR: Unknown variable "sd".
```

**Comment by kakra** [comment]
I think it should use underscores:

    if sd-bus-prover == 'systemd'

**Comment by kakra** [comment]
> Tbh I'm not that happy with combining everything into one option anymore, unless I would rewrite the entire build files so it makes a little bit more sense with this option.

I'm open to a different approach...

**Comment by stephanlachnit** [comment]
> Sorry:

No need to apologize, I currently don't have my normal build system setup so I can't test it^^

**Comment by kakra** [comment]
Here's another one:

    data/meson.build:13:3: ERROR: Unknown variable "sd_bus_prover".

Actually, I also missed to spot that. ;-)

**Comment by stephanlachnit** [comment]
> Here's another one:

At this point it's just sad xd. I'm gonna invest more on time this when I have it, at least I hope all spelling mistakes are now sorted out.

**Comment by kakra** [comment]
Yes, seems so. It doesn't link, tho. I appreciate your investment a lot.

I'll inspect this later, maybe clone your branch and fix what's missing.

No urge to work on this now but here's the problem:
```
# sd_bus_provider = "no-daemon"
daemon/f77b12a@@gamemoded@exe/gamemode-dbus.c.o:gamemode-dbus.c:game_vtable: error: undefined reference to 'sd_bus_object_vtable_format'
daemon/f77b12a@@gamemoded@exe/gamemode-dbus.c.o:gamemode-dbus.c:gamemode_vtable: error: undefined reference to 'sd_bus_object_vtable_format'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by stephanlachnit** [comment]
Looks like it, no idea why though.

**Comment by TheGreatMcPain** [comment]
There's a typo in meson.build on line 165.\
`if sd_bus_provider != 'no-dameon'` should be `if sd_bus_provider != 'no-daemon'`

Other than that, I was able to build it on Gentoo with elogind using [this ebuild.](https://github.com/TheGreatMcPain/TheGreatMcPain-overlay/blob/master/games-util/gamemode/gamemode-9999-r1.ebuild)

**Comment by stephanlachnit** [comment]
> There's a typo in meson.build on line 165.

Ah thanks, that was it.

**Comment by jqadev** [comment]
@stephanlachnit there appeared conflicts in `daemon/meson.build` that I guess prevent @afayaz-feral from doing a review.

**Comment by stephanlachnit** [comment]
> @stephanlachnit there appeared conflicts in daemon/meson.build that I guess prevent @afayaz-feral from doing a review.

Thanks for the heads up, I'll just fix a compiler warning an then it should be ready to be reviewed.

Btw the second commit is optional, I reduces the number of options but it isn't exactly clearer, so any opinions are appreciated.

**Comment by afayaz-feral** [comment]
> Btw the second commit is optional, I reduces the number of options but it isn't exactly clearer, so any opinions are appreciated.

I like that second commit; you wouldn't want multiple options at once so it makes sense to combine them.

**Pull request merged**